### ISSUE TITLE
plot sampling points and density

### ISSUE THREAD

**Issue by Soheyl-Salamikavaki-Bose** [showlab/sparseformer#2]
Hi, 

I want to plot the exact figure of paper (Figure 3 sampling points and density ). But I do not know how I can do it, could you please let me know how I can do it. 
 
Based on your codes I realized that input image with [3, 224, 224] is applied to  self.feat_extractor and gives a tensor with shape [96, 56, 56].
All the process(sampling points and token embedding) is done on this feat with shape [96, 56, 56] not the original image. 
if so, how it is possible to plot figure 3 while the sampling points are done on feat not original image

**Comment by sebgao** [comment]
Hi,
Thank you for your interest in our work. The feature sampling is done on 56x56 feat map but we plot the original input image for better illustration and clarity. Thank you for pointing out the ambiguity and we'll add that note to the caption in the future version.

To plot the original image along with sampling points, you can perform `plt.imshow` and `plt.scatter` continuously on the same plot.

**Comment by Soheyl-Salamikavaki-Bose** [comment]
if you could add that plotting code here, it would be really great.

**Comment by sebgao** [comment]
It is [`sklearn.neighbors.KernelDensity`](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KernelDensity.html) with the following arguments `KernelDensity(bandwidth=15, kernel="tophat")`. Note that `bandwidth=15` is with sampling points in `224x224` raw pixel measurement.

**Comment by sebgao** [comment]
```
def kde2D(x, y, bandwidth=15, xbins=112j, ybins=112j, **kwargs): 
    """Build 2D kernel density estimate (KDE)."""

    # create grid of sample locations
    xx, yy = np.mgrid[0:224:xbins, 
                      0:224:ybins]

    xy_sample = np.vstack([yy.ravel(), xx.ravel()]).T
    xy_train  = np.vstack([y, x]).T

    kde_skl = KernelDensity(bandwidth=bandwidth, kernel='tophat', **kwargs)
    kde_skl.fit(xy_train)

    # score_samples() returns the log-likelihood of the samples
    z = np.exp(kde_skl.score_samples(xy_sample))
    return xx, yy, np.reshape(z, xx.shape)
```

Once you get the results, you can plot them with
```
kdes = kde2D(sampling_x, sampling_y, 15)
plt.pcolormesh(*kdes, alpha=0.7, cmap='jet')
```

Please keep again in mind that the `sampling_x` and `sampling_y` should be scaled to lie in the [0, 224] range. As original sampling coordinates are in the [0, 1] range, you can scale them by multiplying by 224.

**Comment by Soheyl-Salamikavaki-Bose** [comment]
Thank you so much for your response. I have trained the model for different image dataset, they are not is square shape. the sampling values are not between 0 and 1 and density function does not seem to be true.
can you guess what the issue can be, I guess it is related to rectangle shape of images.
![sampling_points_density_map1](https://github.com/showlab/sparseformer/assets/135761038/3c8a05ea-95f0-486b-91d1-40269e3eb613)
![sampling_points_plot](https://github.com/showlab/sparseformer/assets/135761038/78881221-204a-44a9-9126-16be19c21f94)

**Comment by sebgao** [comment]
Hi, sorry for the confusion caused. The sampling points are not bound to be restricted in `[0, 1]` as we do not pose any losses on them. But usually, sampling points generally fall in `[0, 1]` as long as they are *not out of image area*. So your figure look technically right regarding the step of plotting sampling points.

But what I care more is that your own trained model seems to be collapsed (as there are many sampling points out of the image area). Can you tell me the performance of SparseFormer (or comparison to the simple CNN baseline) on your own dataset?

By the way, I think that the training and model code should support non-rect images already but I have not tested it yet.

**Comment by Soheyl-Salamikavaki-Bose** [comment]
The accuracy of the model is good, 90%.
The above sampling is for stage 0. I tried to do sampling for other stages too, but I found out that:
self.layers = [SFUnit, transition_layer, SFUnit, transition_layer] 
where transition_layer is linear layer: token_embedding = transition_layer(token_embedding)


so for stage 0 and 2 I have sampling, but there is no sampling for stage 1 and 3 while in the paper there are 5 stages of sampling. I'm confused. Could you please explain how I can have 5 stage 0 sampling?
